ruma-common/src/push.rs

```diff
@@ -39,12 +39,25 @@ pub struct Ruleset {
     pub underride: Vec<ConditionalPushRule>,
 }
 
+/// An enum that represents a push rule of any shape
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum AnyPushRule {
+    /// A rule with no pattern or conditions
+    Normal(PushRule),
+    /// A rule with a pattern
+    Patterned(PatternedPushRule),
+    /// A rule with conditions
+    Conditional(ConditionalPushRule),
+}
+
 /// A push rule is a single rule that states under what conditions an event should be passed onto a
 /// push gateway and how the notification should be presented.
 ///
 /// These rules are stored on the user's homeserver. They are manually configured by the user, who
 /// can create and view them via the Client/Server API.
 #[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
```

Comment:
Hey, you cheated! :smile: 

More seriously, I don't think this is a good idea. In the Matrix spec, adding a new field is not a breaking change, and I want the same to be true for Ruma. (we're still far away from that being the reality, but I've been working on fixing that, see #14)

Reply:
Oh, I see! Makes sense, but I'm afraid that combined with `pattern` and `conditions` this will make deserialization way trickier to get right.